Create voice note editor

xSabrina opened the pull request:
### **Description**
Closes #106 

A **voice note editor** has been created - therefore noteLists in NotesFragment and BookNotesFragment have been adjusted in code and layout. The editor is only opened on creating a new voice note - existing voice notes in noteLists have small buttons to play/pause, stop and a seekBar to move through the audio.

Also did some general clean ups and changes that affected the VoiceNotes-feature or have been affected by it.

### **Affects**
@SarahKurek should take a look since she created the view for booknotes that has been changed majorly.

### **Notes for Reviewer**

**_1. General changes:_**

- **Adjust and clean up build.gradle:** add necessary implementations for using pulseAnimation and seekBar and set up mindSDK from 24 to 26 (suggested because something from a former merge into the dev is only available from 26 upwards and could cause crashes).
- **adjust and clean up AndroidManifest** by removing legacy usages and adding the necessary one for recording audio.

- **BookNotesView has been renamed to BookNotesFragment** to match the other class name structure.
- **renamed the bookNotesView-xml file** to match the fragment and other xml-name-structure **to "fragment_book_notes"**.
- **BookNotesRecyclerViewAdapter has been removed** and its usages have been **replaced by usages of the NotesRecyclerViewAdapter** to remove much double code.
- **HelpFragment has been removed** and has been **replaced by alert dialogs** with only an positive "OK"-Button to save resources and avoid possible bugs with using the system UI back-button e.g.
- **Reformatted all help text strings** for better readability.
- **fixed a bug with help-strings displays** all over the app (HTML-format has been ignored).
- **Reformatted all deleting text strings** for better readability.
- Some **icon-drawables** have been editted in size (e.g. the microphone for voiceNote-type), necessary ones were added and legacy ones have been removed.
- a **custom seekBarListener** has been created for seekBar-usages.
- **renamed color-layout-xml library_item_selector** to general "item_selector" since we use it for several items.

- did some **small code clean ups** (checkstyle or reformatting) to some classes.
- **fixed a bug when editing text notes** caused by missing bookId param on noteItems. The bookId is now added correctly on creating noteItems in the noteModel by using the noteDao.
- **removed overriding system back button** from TextNoteEditor and VoiceNoteEditor because it was **causing bugs** (on using the bottom nav e.g.) and using the onPause instead (we have been talking about an approve/back-button in our last meeting - if we implement sth like this the onPause code has to be moved to that button onClick).

**_2. VoiceNoteEditorFragment:_**

- on open a necessary **permission to record audio is requested**.
- contains 3 buttons: **stop, record and play(/pause)**. The buttons change their color lightly onClick (select) and the recordButton has a small wave animation beyond it to indicate whenever sound is recorded. The stop-Button works either for recording or playing. The playButton switches to pause on playing audio.
- recorded audio is **saved on leaving the VoiceNoteEditorFragment** (matching the TextNoteEditorFragment). The name is currently default by "Sprachnotiz vom " + current day, month and year.
- contains a **seekBar and small start- and endTime-displays** that are displayed whenever something has been recorded yet (means after the first time using the recordButton): you can move the thumb to any position and play audio from a chosen position.
- contains **optionsMenu** with "Hilfe"-item (necessary help text has been created).

**_3. NotesFragment and BookNotesFragment:_**

- **Layout-xml for BookNotesCardView has been removed** and usages have been replaced by the general one for noteCardView.
- list_view_item_note.xml has been adjusted to **contain a seekBar with start and endTimer and play/pause and stop button**.-
- the **stopButton is only working** when the related playButton of the same listItem was used and the **related audio is playing**.
- on **pressing the playButton** of another listItem **stops** any audio played and **resets** all other voiceNoteItemViews.
- **removed the legacy hiddenDeletePanel** in NotesFragment to be consistent in the app structure.
- **added swipeToDelete to the BookNotesFragmen**t to match the NotesFragment.
- **fixed a small bug** with the noteList expands to **overlap the BottomNavigation in NotesFragment**.
- **use alertDialog and its toast** for the NotesFragment as well to be consistent.

**_4. Database related (NoteModel, NoteDao, etc.):_**

- add **necessary methods** to fetch noteMedia in bookNotesViewModel, noteModel and noteDao.
- adjust **the way notes are created** (constructor params or database handling) to contain byte arrays for putting media on it and save it as BLOB (byte[]) in the noteFileTable.

**_5. Suggestions for your testing:_**

- if you use an emulator in AndroidStudio: open your extended controls and go to "Microphone" and **enable "Virtual microphone uses host audio input"** to use your pc-microphone.

SarahKurek commented on BibBuddy/.idea/jarRepositories.xml:
```diff
@@ -31,5 +31,15 @@
+    </remote-repository>
   </component>
-</project>
+</project>
```
```suggestion
</project>

```

SarahKurek commented on BibBuddy/app/build.gradle:
```diff
@@ -13,7 +16,7 @@ android {
     defaultConfig {
         applicationId "de.bibbuddy"
-        minSdkVersion 24
+        minSdkVersion 26
```
Why not support Android 7 anymore?

xSabrina replied:
Android Studio told me to change it to 26 since we use an Android element that is not supported below and might cause problems. We talked about it in person already though. I changed it back to 24 for our device testing for now.

SarahKurek commented on BibBuddy/app/src/main/res/menu/fragment_voice_editor_menu.xml:
```diff
@@ -0,0 +1,8 @@
+        android:id="@+id/menu_help_voice_editor"
+        android:title="@string/help" />
+
+</menu>
```
```suggestion
</menu>

```

SarahKurek reviewed (changes requested):
Next time please don't do everything on one branch, because now merging with my design changes will be difficult.
Help will be integrated into the bottom navigation bar and should, in my opinion, remain as a separate view. Therefore, please reverse the changes in this regard.

Name changes to classes that are being edited by others at the same time are likely to cause merge problems and should only be made after consultation. So undo the renaming and fix the checkstyle warning please.
Please make changes concerning these issues and then I'll take another look at everything.

xSabrina commented:
> Next time please don't do everything on one branch, because now merging with my design changes will be difficult.

I agree, sorry for now though. I merged conflicts manually.

> Help will be integrated into the bottom navigation bar and should, in my opinion, remain as a separate view. Therefore, please reverse the changes in this regard.

We talked about this in our meeting already - so I will leave the changes.

> Name changes to classes that are being edited by others at the same time are likely to cause merge problems and should only be made after consultation. So undo the renaming and fix the checkstyle warning please.

True - there is no undoing necessary though, I fixed possible merge conflicts already.

xSabrina commented:
I solved all merge conflicts and fixed some bugs.

Mainly fixing the bug from #119:
Also limited the possible recording duration for voice notes to 1 Min. to avoid crashes from too large byte arrays for our database blobs. I chose the limit in time and not file size, because it is more resource saving to use a countDown instead of permanently checking the temporary-recording-file for its size. I chose 1 minute as limit since it is almost the biggest possible size and a plain value. If the record extends over 1 Min. the audio is saved as a separate note (a toast to notify the user additionally) and automatically starts the next recording.
Also adjusted the help strings for the voice note editor to this.

SarahKurek commented on BibBuddy/app/src/main/res/layout/fragment_book_notes.xml:
```diff
@@ -1,58 +1,135 @@
 
-    <androidx.recyclerview.widget.RecyclerView
-        android:id="@+id/book_notes_recycler_view"
-        android:layout_width="match_parent"
-        android:layout_height="wrap_content"
-        android:layout_marginTop="58dp"
-        android:clipToPadding="false"
-        android:orientation="vertical"
-        android:scrollbars="vertical"
-        app:layoutManager="LinearLayoutManager" />
-
-    <com.google.android.material.floatingactionbutton.FloatingActionButton
-        android:id="@+id/btn_add_note"
+        <Button
+            android:id="@+id/btn_add_note"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_below="@id/note_name"
+            android:layout_marginStart="20dp"
+            android:layout_toEndOf="@id/btn_back"
+            android:text="@string/add_note"
+            android:theme="@style/Button.Pink" />
+        -->
```
Comments should be deleted.

xSabrina replied:
True - deleted.

SarahKurek commented on BibBuddy/app/src/main/res/layout/fragment_book_notes.xml:
```diff
@@ -1,58 +1,135 @@
 <?xml version="1.0" encoding="utf-8"?>
-
-<androidx.coordinatorlayout.widget.CoordinatorLayout xmlns:android="http://schemas.android.com/apk/res/android"
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Why change to Relative Layout? Now you cannot scroll to the button.

xSabrina replied:
There was an error and I thought it might have been caused by the layout - you are right though, I changed that back.

SarahKurek commented on BibBuddy/app/src/main/res/layout-v26/fragment_book_notes.xml:
```diff
@@ -0,0 +1,158 @@
+<?xml version="1.0" encoding="utf-8"?>
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:id="@+id/fragment_book_notes"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:layout_marginTop="65dp">
```
Why fragment_book_note.xml as layout and as laayout-v26?

xSabrina replied:
AndroidStudio created that by itself after changing minSDK to 26.

SarahKurek commented on BibBuddy/app/src/main/java/de/bibbuddy/NoteModel.java:
```diff
@@ -47,8 +64,43 @@ public Note getLastNote() {
+   * @return returns a string in "dd.mm.yyyy hh:mm:ss"-format.
+   */
+  public String getDate(Long date) {
+    Date d = new Date(date);
+    SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX",
+        Locale.getDefault());
+
+    String string = simpleDateFormat.format(d);
+    String day = string.substring(8, 10);
+    String month = string.substring(5, 7);
+    String year = string.substring(0, 4);
+    String time = string.substring(11, 16);
+    string = day + "." + month + "." + year + " " + time + " Uhr";
```
Use noteItem.getModDateString with the adapter (line 134 in NotesRecyclerViewAdapter).

xSabrina replied:
Thanks - using the new suggested method now.

SarahKurek commented on BibBuddy/app/src/main/java/de/bibbuddy/NoteDao.java:
```diff
@@ -223,23 +226,44 @@ public String findTextById(Long id) {
+        DatabaseHelper._ID + "=?", new String[] {String.valueOf(noteFileId)},
+        null, null, null, null);
+
+    cursor.moveToFirst();
+    byte[] bytes = cursor.getBlob(1);
+    cursor.close();
```
try ... catch

LuisMossburger commented:
As discussed in team meeting, this PR will be closed as further major changes are needed.
